Ranking System v2

francois-rozet opened the pull request:
Hello @anuraghazra :wave:

As #883, #455 and #1029, I've noticed that the function `calculateRank` doesn't work as expected.

The first problem is the use of a normal distribution instead of an [exponential distribution](https://en.wikipedia.org/wiki/Exponential_distribution), as pointed out by #960 . The second is that each metric has the same weight in the score, while some are clearly better indicators (e.g. stars). I've corrected both problems in this PR.

If you want to try it out before merging, I've hosted this version on Vercel.

![](https://github-readme-stats-francois-rozet.vercel.app/api?username=torvalds)

![](https://github-readme-stats-francois-rozet.vercel.app/api?username=anuraghazra)

![](https://github-readme-stats-francois-rozet.vercel.app/api?username=francois-rozet)

![](https://github-readme-stats-francois-rozet.vercel.app/api?username=meurissemax)

Importantly, in #960, being extremely good in a single metric (e.g. 100k commits) while being terrible elsewhere still gets an "S+" rank. In my implementation, in order to get "S+", you need to be outstanding everywhere. However, you can still get "A+" or "S" if a single metric is poor (e.g. contributions or issues) but you have a lot of stars and followers.

anuraghazra commented:
So in #960 seems like linustorvals is getting S+ rank which is good, since atleast we know a baseline that there are people who can get S+ but in this PR your changes seems to stretch the stats maybe a little bit too far?

francois-rozet commented:
Hi @anuraghazra 👋 I've edited the weights such that Linus Torvalds is now S+.

By the way S+ is top 2.5%, S 10%, A+ 25%, A 50% and B+ 75% and B 100%.

RodrigoDornelles commented:
> o in #960 seems like linustorvals is getting S+ rank which is good, since atleast we know a baseline that there are people who can get S+ but in this PR your changes seems to stretch the stats maybe a little bit too far?

thought it was great, in the current algorithm most are in the A+, even accounts with less than 5 commits and no stars get this rank, due to the huge amount of registered guests. and accounts with numerous commits, repositories and pull requests also remain with the same rank as this one.

anuraghazra commented:
> By the way S+ is top 2.5%, S 10%, A+ 25%, A 50% and B+ 75% and B 100%.

This ratio seems well balanced.

francois-rozet commented:
@anuraghazra Any updates on you plans ?

Enforcer007 commented:
This is one of the good ways of modeling the data to exponential distribution. I frequently work with datasets and most of them don't fit into normal distribution (I think the current ranking depends on fitting the data to a normal distribution). 

As we currently have no idea on the whole data i.e, meta-data for all the github users with respect to commits, prs, stargazers etc..,. It would be difficult to obtain right parameter values and weights for the model to explain the distribution,

For fun I've tried to see what's the minimum count(MC) required for only one type of parameter (considering others to be zero count) to achieve a given rank.

|Rank|threshold|MC - Commits|MC - Prs|MC-Followers|MC-Stars|
|---|---|---|---|---|---|
|S+|2.5%|None|None|None|None|
|S+|10%|None|None|None|None|
|A+|25%|None|None|None|None|
|A|50%|None|None|None|300|
|B+|75%|None|None|None|83|
|B|100%|None|None|None|None|

Interestingly the weights are so strict that one can achieve only upto Rank A who has only a minimum of 300 stargazers.

I'll try to share more stats based on combinations of parameters.

francois-rozet commented:
Hello @Enforcer007, I'm not sure to understand what you want to say.

Do you think it is bad that one cannot get S+ (or even A+ in fact) by maximizing a single metric ?

Enforcer007 commented:
Hi @francois-rozet,

The weights you used is a good start, as it restricts users to go higher in ranks who only focus on one kind type of parameter.

anuraghazra commented:
Just testing out few numbers with the ranking system 

https://codesandbox.io/s/calculaterank-9nj6d?file=/src/index.js


One scenario which I'm seeing is that if someone is overall decent on all stats but has 0 stars gained they will always get B+ rank which isn't ideal. 

See codesandbox for more info.

francois-rozet commented:
Hello @anuraghazra,

You seem to have a really warped view of what is a "decent" number of stars. Most users don't have any stars, let alone more than 100. "just a few thousand stars" is already incredibly outstanding. For instance, [here](https://gitstar-ranking.com/users?page=50) you see that only 5000 users have more than 2750 stars and there are over 40 millions active users on GitHub.

Same for the number of commits, followers, PRs, etc. (2500 commits or 750 followers is not average at all). The values I have chosen were calculated (by hand) as the averages over the stargazers of one of my repos ([francois-rozet/piqa](https://github.com/francois-rozet/piqa)). The stargazers of this repository (there are ~30k) would be much more representative. Since you are familiar with GitHub's API it shouldn't be too hard for you to compute gather the data of these users and then compute the averages.

anuraghazra commented:
> You seem to have a really warped view of what is a "decent" number of stars. Most users don't have any stars, let alone more than 100

Exactly thats why I said why  if someone is overall decent on all stats but has 0 stars gained they will always get B+ rank which is really unfair. 

I'm talking about the first case 
```

// Average user but with 0 stars
console.log(
  calculateRank({
    totalCommits: 1500,
    followers: 500,
    prs: 150,
    issues: 65,
    stargazers: 0
  })
);
```

francois-rozet commented:
You can reduce the weight of stars if you want. `STARS_WEIGHT = 0.5` seems balanced.

rickstaa commented:
@francois-rozet Thanks a lot for your pull request! I however think the rank calculation should also take the number of contributions into account (see https://github.com/anuraghazra/github-readme-stats/pull/960#issuecomment-963171112).

rickstaa commented on tests/calculateRank.test.js:
```diff
@@ -2,17 +2,45 @@ require("@testing-library/jest-dom");
+        issues: 13,
+        stargazers: 100,
+      }),
+    ).toStrictEqual({"A", score: 50.});
```
```suggestion
    ).toStrictEqual({level: "A", score: 50.});
```

Otherwise, the tests will throw a `Unexpected token (30:24)` error.

rickstaa commented:
@francois-rozet I reviewed your pull request and found one small bug in the tests. When that is fixed, I think it should be fine to be merged into the master branch.

markus-wa commented:
any updates on this?

francois-rozet commented:
Any updates @anuraghazra ?

francois-rozet commented:
Hello @anuraghazra, I have made this PR **more than one year** ago. To be honest, I find it disrespectful that you ignore all PRs fixing the rank calculation, both for PR authors and users who have been asking you to fix this problem for years. If you don't want to change the rank calculation, just say it on the README, but don't let people put efforts into fixing it for nothing.

If you do consider changing the rank calculation, **please make a move**.

There are only two PRs that propose good systems: #960 and #1186. They are similar, but in #960, being extremely good in a single metric (e.g. 100k commits) while being terrible elsewhere still gets an "S+" rank. In #1186, in order to get "S+", you need to be outstanding everywhere. However, you can still get "A+" or "S" if a single metric is poor (e.g. contributions or issues) but you have a lot of stars and followers.

I feel that #1186 is more balanced and both @rickstaa and @GaryStriving have approved the changes.

rickstaa commented:
@francois-rozet, I understand your frustration. Unfortunately, @anuraghazra and I maintain this repo in our free time, so merging PR might go slightly slower than in a company repo. I, however, understand that one year is a bit too long. 😅 I think my proposal in #1867 will speed up the merging process since it will allow us to include more maintainers since an organization gives more detailed permission control.

Having that said. The main difficulty with your PR is that #960 also addresses the issue you are addressing. As stated in the past, my preference is with #1186, but @anuraghazra hasn't yet stated an opinion. I am happy to merge #1186, but I would love a second opinion from @anuraghazra since it regards the core codebase. Maybe let's give @anurra 30 days to respond, after which I will merge it. 🤔  We can always revert it back to the old version.

rickstaa commented:
@francois-rozet Since last month, we required prettier to be run on each PR (see https://github.com/anuraghazra/github-readme-stats/actions/runs/3111026424/jobs/5042849646). I created https://github.com/francois-rozet/github-readme-stats/pull/2 to fix the errors thrown by the prettier check. You can, of course, also run prettier yourself. Whatever you prefer. 🤓

francois-rozet commented:
I merged your PR.

francois-rozet commented:
Hello @rickstaa, it has been a bit more than a month. Is it finally time to merge? 🙏

FleetAdmiralJakob commented:
Pls. merge

rickstaa commented:
> Hello @rickstaa, it has been a bit more than a month. Is it finally time to merge? pray

@francois-rozet I think we can merge this if we set `STARS_WEIGHTS` to `0.75` or `0.5`. @anuraghazra, what do you think?

rickstaa commented:
@francois-rozet since this is one of our most requested features, I think it is essential it gets merged (see https://github.com/anuraghazra/github-readme-stats/issues/1935). Sadly, it affects everything that uses the public instance, so I would really like @anuraghazra opinion on it.

dreamyguy commented:
On https://github.com/anuraghazra/github-readme-stats/pull/960#issuecomment-914357213 (Sep 7, 2021) @anuraghazra wrote:

> Will consider both of the PRs, and release it under experimental flags ?enable_experiments=new_ranking_system to get feedback from the community and roll out the change gradually.

This would not affect production as it would be offered through an experimental flag.

I maintain open-source projects and know it's annoying to get bugged on un-merged PRs but it's been **549 days**: 1.5 years to the day [and counting](https://dreamyguy.github.io/moments/#/1/relative-to-now/date-a/2021/July/13/22/53/53/).

क्यों 🤷

rickstaa commented:
> On [#960 (comment)](https://github.com/anuraghazra/github-readme-stats/pull/960#issuecomment-914357213) (Sep 7, 2021) @anuraghazra wrote:
> 
> > Will consider both of the PRs, and release it under experimental flags ?enable_experiments=new_ranking_system to get feedback from the community and roll out the change gradually.
> 
> This would not affect production as it would be offered through an experimental flag.
> 
> I maintain open-source projects and know it's annoying to get bugged on un-merged PRs but it's been **549 days**: 1.5 years to the day [and counting](https://dreamyguy.github.io/moments/#/1/relative-to-now/date-a/2021/July/13/22/53/53/).
> 
> क्यों 🤷

My preferences keeps going to this pull request instead. It is more balanced. I however still think star weight has to be updated. I didn't merge it since I wanted one last opinion of @anuraghazra.

FleetAdmiralJakob commented:
@anuraghazra Can you merge?

rickstaa commented:
I just rebased this PR and changed the star weight to 0.75. Although the ranking is probably not to everybody's liking, I think it is a huge improvement over the current behaviour:

```md
![](https://github-readme-stats-git-patch-rank-rickstaa.vercel.app/api?username=torvalds)
```

![](https://github-readme-stats-git-patch-rank-rickstaa.vercel.app/api?username=torvalds)

````md
![](https://github-readme-stats-git-patch-rank-rickstaa.vercel.app/api?username=anuraghazra)
```

![](https://github-readme-stats-git-patch-rank-rickstaa.vercel.app/api?username=anuraghazra)

````md
![](https://github-readme-stats-git-patch-rank-rickstaa.vercel.app/api?username=francois-rozet)
```

![](https://github-readme-stats-git-patch-rank-rickstaa.vercel.app/api?username=francois-rozet)

````md
![](https://github-readme-stats-git-patch-rank-rickstaa.vercel.app/api?username=meurissemax)
```

![](https://github-readme-stats-git-patch-rank-rickstaa.vercel.app/api?username=meurissemax)
````md
![](https://github-readme-stats-git-patch-rank-rickstaa.vercel.app/api?username=meurissemax)
```

![](https://github-readme-stats-git-patch-rank-rickstaa.vercel.app/api?username=meurissemax)

rickstaa commented:
@anuraghazra I think I will merge this in 20 days as we can always revert to the previous behaviour. You are welcome to let me know if you want changes before that. This issue has been open for a long time, and I think the solution proposed by @francois-rozet is a great improvement 🚀.

guylepage3 commented:
@rickstaa stars are just a form of bookmarking for most developers. Commits should be weighted the same or more than the stars. I respect real developers. Not marketers.

francois-rozet commented:
> @rickstaa stars are just a form of bookmarking for most developers. Commits should be weighted the same or more than the stars. I respect real developers. Not marketers.

IMO, the number of commits is almost meaningless. You can be a great dev with few, well thought commits. Stars, PRs and issues are much better metrics of impact.

BTW @rickstaa, we could increase the weight of PRs and Issues to 0.5, but users that don't use the PR/Issue system of GitHub, like Linus Torvalds, will drop in the ranking.

guylepage3 commented:
I disagree. This is a bit anecdotal, but my experience building a lot of software used by many has been the opposite. This also goes the same for almost all the major devs I know who have built software such as bootstrap, YC companies, etc.

rickstaa commented:
> @rickstaa stars are just a form of bookmarking for most developers. Commits should be weighted the same or more than the stars. I respect real developers. Not marketers.

Hey @guylepage3. Thanks for your comment. I think you misunderstood my intention here 😅. I, personally, do not care about the rank on the stats card (see https://github.com/rickstaa). I just wanted to improve the current ranking and close one of our most requested features (see https://github.com/anuraghazra/github-readme-stats/issues/1935).

Personally, I think ranking developers on this limited number of parameters is very hard and subject to discussion. I for example, would give the highest weight to the pull request statistic, but others might rate the stars as more important. The only thing I did in this pull request was lowering the stars' weight in this pull request from `1.0` to `0.75` to make the ranking a bit more weighted in my opinion. You are welcome to give suggestions for the weights which you think would give reasonable results 🤔. 

#### Another ranking scheme

As an example, I quickly created another ranking scheme which, I think, would make sense in my opinion. In this scheme, more weight is given to commits and pull requests. It, however, will give https://github.com/torvalds a `A+` ranking, which many people will dislike. It further is a bit steep, but this can be solved by changing the means.

```md
![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=torvalds)
```

![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=torvalds)

```md
![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=rickstaa)
```

![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=rickstaa)

```md
![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=anuraghazra)
```

![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=anuraghazra)

```md
![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=francois-rozet)
```

![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=francois-rozet)

```md
![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=FleetAdmiralJakob)
```

![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=FleetAdmiralJakob)

```md
![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=meurissemax)
```

![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=meurissemax)

### Normal & include_all_commits=true

```md
![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=torvalds&include_all_commits=true)
```

![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=torvalds&include_all_commits=true)

```md
![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=anuraghazra&include_all_commits=true)
```

![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=anuraghazra&include_all_commits=true)

```md
![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=rickstaa&include_all_commits=true)
```

![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=rickstaa&include_all_commits=true)

```md
![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=francois-rozet&include_all_commits=true)
```

![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=francois-rozet&include_all_commits=true)

```md
![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=FleetAdmiralJakob&include_all_commits=true)
```

![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=FleetAdmiralJakob&include_all_commits=true)

```md
![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=meurissemax&include_all_commits=true)
```

![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=meurissemax&include_all_commits=true)

rickstaa commented:
> > @rickstaa stars are just a form of bookmarking for most developers. Commits should be weighted the same or more than the stars. I respect real developers. Not marketers.
> 
> IMO, the number of commits is almost meaningless. You can be a great dev with few, well thought commits. Stars, PRs and issues are much better metrics of impact.
> 
> BTW @rickstaa, we could increase the weight of PRs and Issues to 0.5, but users that don't use the PR/Issue system of GitHub, like Linus Torvalds, will drop in the ranking.

@francois-rozet thanks for jumping in. I didn't see your comment when I wrote https://github.com/anuraghazra/github-readme-stats/pull/1186#issuecomment-1500955729. I now understand why you did your weights like that. I also quickly tried to test out some weights:

https://github.com/anuraghazra/github-readme-stats/blob/57b1d8363975627091346dae33e98759410e594b/src/calculateRank.js#L14-L44

I, however, noticed it is very hard to keep people like [Linus Torvalds](https://github.com/torvalds) with an S+ rank when they don't use the classical GitHub approach. I sadly could not find a good middle ground yet. Further, I accidentally gave myself a `S+` rank, which I don't agree with 🤣. I therefore still think your ranking scheme is more weighted 👍🏻. But like I said, this discussion is a bit biased and the ranking will never make everybody happy. I, however, think that the current ranking mechanism in which users start at A+ and have a very hard time getting to S or even S+ is far from ideal 🤔.

## Another ranking scheme

### Yearly commits

```md
![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=torvalds)
```

![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=torvalds)

```md
![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=rickstaa)
```

![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=rickstaa)

```md
![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=anuraghazra)
```

![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=anuraghazra)

```md
![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=francois-rozet)
```

![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=francois-rozet)

```md
![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=FleetAdmiralJakob)
```

![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=FleetAdmiralJakob)

```md
![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=meurissemax)
```

![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=meurissemax)

### Weight 0.75 & include_all_commits=true

```md
![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=torvalds&include_all_commits=true)
```

![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=torvalds&include_all_commits=true)

```md
![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=anuraghazra&include_all_commits=true)
```

![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=anuraghazra&include_all_commits=true)

```md
![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=rickstaa&include_all_commits=true)
```

![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=rickstaa&include_all_commits=true)

```md
![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=francois-rozet&include_all_commits=true)
```

![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=francois-rozet&include_all_commits=true)

```md
![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=FleetAdmiralJakob&include_all_commits=true)
```

![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=FleetAdmiralJakob&include_all_commits=true)

```md
![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=meurissemax&include_all_commits=true)
```

![](https://github-readme-stats-git-patch-rank-rstaa-rickstaa.vercel.app/api?username=meurissemax&include_all_commits=true)

chrisK824 commented:
Hey guys, my two cents here is that a ranking system should have more degrees so that people have goals to give them motivation. That said I would say that people should start from lower ranks and logarithmicly ramp up the ladder. I see no point having everyone stuck on A or A+ forever

rickstaa commented:
> Hey guys, my two cents here is that a ranking system should have more degrees so that people have goals to give them motivation. That said I would say that people should start from lower ranks and logarithmicly ramp up the ladder. I see no point having everyone stuck on A or A+ forever

@chrisK824, thank you very much for your two cents. I think you "Hit the nail on the head". The current ranking is not very motivational for new users. People have been complaining about this for a long time. I think @francois-rozet is the best ranking we have seen so far, and I am in favour of merging it. People can then always can create new pull requests to improve it further. @anuraghazra and @Zo-Bro-23 what do you think?

@francois-rozet the only thing that, I think, would improve your ranking further would be:
- To split the `PRS_ISSUES_WEIGHT ` into two separate weights. This would make it easier for people to suggest improvements in the future. I, however, am not sure if it is easy to get similar ranking when these items are split into separate weights.

https://github.com/anuraghazra/github-readme-stats/blob/9900d74f3e5c7ae0e805035f878c08f9999e0d93/src/calculateRank.js#L29

francois-rozet commented:
> To split the PRS_ISSUES_WEIGHT into two separate weights. This would make it easier for people to suggest improvements in the future. I, however, am not sure if it is easy to get similar ranking when these items are split into separate weights.

The reason why I bundled PRs and Issues is because the number of issues alone are not a good target. You can open 10k issues without ever writing code. You can also fix a ton of issues without opening any. If we do split them, issues must have a very low target (`ISSUES_MEAN`) and/or a very low weight. So low that we could actually drop it altogether.

As a side note if `X = X_MEAN`, the contribution of `X` to the rank is `0.5 * X_WEIGHT`. To get a `0.9 * X_WEIGHT` contribution, you need `X > 4 * X_MEAN`. So looking at the current settings, I think that

1. The commit target is too high (4 * 500 = 2000 commits in a single year is enormous). We could drop it to 250.
2. The PRs target is a bit low (4 * 25 = 100 PRs is not that much). We could increase it to 50, and also increase the weight.
3. I would drop the number of issues altogether.

So something like

```javascript
  const COMMITS_MEAN = 250,
    COMMITS_WEIGHT = 0.25;
  const FOLLOWERS_MEAN = 50,
    FOLLOWERS_WEIGHT = 0.25;
  const PRS_MEAN = 50,
    PRS_WEIGHT = 0.5;
  const STARS_MEAN = 100,
    STARS_WEIGHT = 0.75;
```

Tell me what you think!

chrisK824 commented:
> By the way S+ is top 2.5%, S 10%, A+ 25%, A 50% and B+ 75% and B 100%

I think that those ranks should be more @rickstaa @francois-rozet. Extending this to 10% clusters would be more motivational, even though new users get a low degree it still is better.

e.g E=100%, E+=90%, D=80%, D+=70%, C=60%, C+=50%, B=40%, B+=30%, A=20%, A+=10%, A++ or S = 2.5%

or even better, just drop all the cluster system and present exact percentage % for each user, so that people can see their contributing effect daily/weekly and get motivated.

francois-rozet commented:
Hello @chrisK824, I think this PR is already a big improvement.  Further changes should be discussed later.

rickstaa commented:
> I disagree. Your number of PRs is huge! At the very least an `S`  If we raise the bar such that you don't get an `S`, no one will...

Thanks for the kind words 🙏🏻.

I think the changes you suggest will create a rank calculation that is both balanced and motivational. I am happy to merge your PR if they are implemented 👍🏻. I, however, had some small comments:

-  Although 2k commits per year is doable, I agree that it is a bit optimistic. I, think the change you suggest makes sense. We, however, have to then make the rank dependent on whether the user has `include_all_commits` enabled.
- I would, include the `ISSUES` item with a very low weighting in the code. This, however, is a subjective and somewhat puristic choice, so I would also be good with omitting it.

rickstaa commented:
> Hello @chrisK824, I think this PR is already a big improvement. Further changes should be discussed later.

Good suggestion. I think we can tackle this in a successive pull request 👍🏻.

francois-rozet commented:
> We, however, have to then make the rank dependent on whether the user has `include_all_commits` enabled.

Either we modify the threshold when `include_all_commits` is enabled/disabled, or we always use the total number of commits (make the rank invariant with respect to `include_all_commits`). I prefer the latter.

rickstaa commented:
> > We, however, have to then make the rank dependent on whether the user has `include_all_commits` enabled.
> 
> Either we modify the threshold when `include_all_commits` is enabled/disabled, or we always use the total number of commits (make the rank invariant with respect to `include_all_commits`). I prefer the latter as it means the rank can only increase with time.

I agree that calculating the rank based on the total commit count makes more sense. However, since we use the REST API to fetch the total commit count and the GraphQL API for the year commit count, this will lead to two fetch requests instead of one when users have `include_all_commits` set to `false`. This could, therefore, unfortunately lead to longer load times and possible trigger GitHub or Vercel timeouts. This therefore has to be tested first 🤔. 

@Zo-Bro-23, @anuraghazra what are your thoughts on this matter?

francois-rozet commented:
Isn't it possible to fetch both the current and total number of commits as one request? If this is too complex, or requires many changes to the codebase, let's do that later and use a high target for the commits (anyway most people include the total number of commits).

dreamyguy commented:
Liking to see the burst of communication, which will hopefully lead to a merge. ✨ 

I feel these changes are among the most requested ones, and it's going to be great to see it live, as it is, for now.

A few points on commit count, though:

- Commits, by themselves, _do not reflect **code quality** nor **skill**, no matter how one looks at it.
- When contributing to high-stake/widely-shared projects, commits are usually squashed into a single commit.
- It can be automated. Google "Github Contributions Graph Art".

**Merged pull-requests**, on the other hand, denote contributions that are likely to be peer-reviewed.
**Issues** require some focus and is time-consuming. Is there anything like **Resolved issues**? That would be a better than just "issue".

Github itself has 3 achievements that hints extra dedication:

- **Starstruck** (star count)
- **Pair Extraordinaire** (co-authored pull-request commits)
- **Pull Shark** (number of merged PRs)

Maybe consider weighing these some more?

Stars, for me, are perhaps the most important metric. Most devs I know are very picky when starring repos, and some even keep their starred repos in check, pruning what's no longer relevant. I for one used to be more generous on giving stars, not because I don't mean to, but because there's just so much out there. I still use it, but the bar is constantly being raised.

chrisK824 commented:
I'll agree with people commented before, stars are important as they demonstrate how useful someone's work has proven for others and recognised in the community.

So @guylepage3 , I wouldn't say it is for marketing. Although I use the feature as bookmark as well, it's a good way to count usability from others in the community.

That said, of course commits are important here as well, showing the work put in by someone.

IMO both should weigh quite a lot in ranking. 

As per stars, this isn't something to earn easy - as time goes by it is harder and harder to implement something that isn't done already in a good/better way. For that reason maybe the stars should play a role in weighting more and more as they stack, as this is showcase of building something that is popular out there. 

(I have 0 stars, so just a fresh view on the topic hehe)

rickstaa commented:
> Isn't it possible to fetch both the current and total number of commits as one request? If this is too complex, or requires many changes to the codebase, let's do that later and use a high target for the commits (anyway most people include the total number of commits).
> 
> Edit: I took a look and it's not possible in a single fetch. Also it seems like the `count_private` option adds the `restrictedContributionsCount` to the number commits, even though it is not a number of commits.
> 
> https://github.com/anuraghazra/github-readme-stats/blob/879937c11d08154f9afa0de7dab7b4945e452276/src/fetchers/stats-fetcher.js#L245-L248

We can or test it out to see if GitHub's time limits are hit or, for now, use a different mean for when people enable `include_all_commits`. We can change this behaviour later when we release the GitHub action version where those problems do not occur 👍🏻.

> Edit: I took a look and it's not possible in a single fetch. Also it seems like the `count_private` option adds the `restrictedContributionsCount` to the number commits, even though it is not a number of commits.

Your correct the naming is wrong (see https://github.com/anuraghazra/github-readme-stats/issues/2517).

rickstaa commented:
> Liking to see the burst of communication, which will hopefully lead to a merge. ✨
> 
> I feel these changes are among the most requested ones, and it's going to be great to see it live, as it is, for now.
> 
> A few points on commit count, though:
> 
> * Commits, by themselves, _do not reflect_ **code quality** nor **skill**, no matter how one looks at it.
> * When contributing to high-stake/widely-shared projects, commits are usually squashed into a single commit.
> * It can be automated. Google "Github Contributions Graph Art".
> 
> **Merged pull-requests**, on the other hand, denote contributions that are likely to be peer-reviewed. **Issues** require some focus and is time-consuming. Is there anything like **Resolved issues**? That would be a better than just "issue".
> 
> Github itself has 3 achievements that hints extra dedication:
> 
> * **Starstruck** (star count)
> * **Pair Extraordinaire** (co-authored pull-request commits)
> * **Pull Shark** (number of merged PRs)
> 
> Maybe consider weighing these some more?
> 
> Stars, for me, are perhaps the most important metric. Most devs I know are very picky when starring repos, and some even keep their starred repos in check, pruning what's no longer relevant. I for one used to be more generous on giving stars, not because I don't mean to, but because there's just so much out there. I still use it, but the bar is constantly being raised.

Rest assured. I will merge this pull request in one form or another 👍🏻. I rather improve the ranking in successive pull requests than keep the terrible ranking system we currently have 😅. Furthermore, I agree with you. I think the most important is stars, then pull requests, then commits and then the rest. The weights and means should reflect that.

rickstaa commented:
@francois-rozet, maybe for now let's implement the changes you suggested above while creating two means for the `include_all_commits` setting? Following, I think it would be nice to include the issue item as a separate item but with a very low weight. After those changes, I will then do one last review of how it influences the ranking and merge it. 

In the future, we could then tune the weights and means even further if people create issues about it. We could even create a small website where people can use sliders to check which ranking they would have with certain stats (see https://github.com/anuraghazra/github-readme-stats/issues/1761). Helping us to discover issues faster.

As stated above, I think your ranking mechanism is a great improvement compared to the old mechanism, and it solves a greatly overdue feature request. I think with time, the weights and mean values will find an equilibrium 🙏🏻.

francois-rozet commented:
Hello @rickstaa, I added `include_all_commits` to the function and modified the weights. Could you check if the new ranking makes sense ? I could not host this version on Vercel.

FleetAdmiralJakob commented:
Hey, I was just curious about what is with A++ and B++ in this Ranking System

francois-rozet commented:
They don't exist. See https://github.com/anuraghazra/github-readme-stats/pull/1186#issuecomment-882058338

rickstaa commented on src/calculateRank.js:
```diff
@@ -1,105 +1,64 @@
+}) {
+  const COMMITS_MEAN = all_commits ? 500 : 100, COMMITS_WEIGHT = 1;
+  const PRS_MEAN = 50, PRS_WEIGHT = 2;
+  const ISSUES_MEAN = 10, ISSUES_WEIGHT = 1;
```
Issues now has the same weight as commits 🤔?

francois-rozet replied:
Fixed ;)

rickstaa reviewed (changes requested):
The code looks very clean, thanks! I had one remark about the issue's weight. Apart from that, it looks like some of the stats fetcher and API tests broke. I will check those later, but you can also take a look if you like.

francois-rozet commented:
I fixed the issues by removing all mentions to `count_private` (even in the README). We can add it back at any time when the GitHub API changes. Feel free to modify the PR all you want, @rickstaa.

chrisK824 commented:
hmm so private contributions are thrown from now on? Time to merge @rickstaa ?

rickstaa commented:
> hmm so private contributions are thrown from now on? Time to merge @rickstaa ?

No, I will take a second look. The tests probably have to be updated. Will review this pull request soon after some deadlines 👍🏻.

rickstaa commented:
> > hmm so private contributions are thrown from now on? Time to merge @rickstaa ?
> 
> No, I will take a second look. The tests probably have to be updated. Will review this pull request soon after some deadlines 👍🏻.

@francois-rozet can you please re-add `counte_private`. Removing a parameter to fix tests is a bit overkill. Although we might remove it in the future (see https://github.com/anuraghazra/github-readme-stats/issues/2517) currently no decision has been taken on this. The Mocking of the data in the tests should be fixed instead.

chrisK824 commented:
Has this been frozen once again? 😋

rickstaa commented:
> Has this been frozen once again? yum

No, I will merge when @francois-rozet makes the changes I requested 🚀. However, let's not pressure him, we are not in a rush 👍🏻.

francois-rozet commented:
Don't worry i'll do it, just in rush for some deadlines 😅

rickstaa commented:
> Don't worry i'll do it, just in rush for some deadlines sweat_smile

I know the feeling. Good luck. No rush from my side 🚀!

chrisK824 commented:
Ahh I see, cool cool, take your time mate, cheers!

francois-rozet commented:
I added `count_private` back. Note that it does not have an effect on the card anymore, so I removed the corresponding section in the README. In the future `count_private` should affect the number of contributions, not of commits. We could consider replacing the `commits` stat by `contributions` as the former actually contributes to the latter (contributions = commits + issues + prs). @rickstaa could you generate cards with the new weights, for a last check before merging? Cheers!

rickstaa commented:
> I added `count_private` back. Note that it does not have an effect on the card anymore, so I removed the corresponding section in the README. In the future `count_private` should affect the number of contributions, not of commits. We could consider replacing the `commits` stat by `contributions` as the former actually contributes to the latter (contributions = commits + issues + prs). @rickstaa could you generate cards with the new weights, for a last check before merging? Cheers!

@francois-rozet I fully support your decision to remove the `count_private` parameter since it is confusing and does not do what the documentation says it does (i.e. #2517). However, I think this should be done in a separate pull request so that the other core collaborators' (i.e. @anuraghazra, @Zo-Bro-23 and @qwerty541) can also look at it. I therefore enabled it again in 0917e03fbba3c500305366aef0cc7ed710208da1. You are welcome to create this pull request 👍🏻.

rickstaa commented:
@francois-rozet thanks for applying my requested changes. Here is the new ranking behaviour. I will let this pull request open for 7 days so that people have time to state their concerns:


## Normal

```md
![](https://github-readme-stats-patch-rank.vercel.app/api?username=torvalds)
```

![](https://github-readme-stats-patch-rank.vercel.app/api?username=torvalds)

```md
![](https://github-readme-stats-patch-rank.vercel.app/api?username=rickstaa)
```

![](https://github-readme-stats-patch-rank.vercel.app/api?username=rickstaa)

```md
![](https://github-readme-stats-patch-rank.vercel.app/api?username=anuraghazra)
```

![](https://github-readme-stats-patch-rank.vercel.app/api?username=anuraghazra)

```md
![](https://github-readme-stats-patch-rank.vercel.app/api?username=francois-rozet)
```

![](https://github-readme-stats-patch-rank.vercel.app/api?username=francois-rozet)

```md
![](https://github-readme-stats-patch-rank.vercel.app/api?username=FleetAdmiralJakob)
```

![](https://github-readme-stats-patch-rank.vercel.app/api?username=FleetAdmiralJakob)

```md
![](https://github-readme-stats-patch-rank.vercel.app/api?username=meurissemax)
```

![](https://github-readme-stats-patch-rank.vercel.app/api?username=meurissemax)

## include_all_commits=true

```md
![](https://github-readme-stats-patch-rank.vercel.app/api?username=torvalds&include_all_commits=true)
```

![](https://github-readme-stats-patch-rank.vercel.app/api?username=torvalds&include_all_commits=true)

```md
![](https://github-readme-stats-patch-rank.vercel.app/api?username=anuraghazra&include_all_commits=true)
```

![](https://github-readme-stats-patch-rank.vercel.app/api?username=anuraghazra&include_all_commits=true)

```md
![](https://github-readme-stats-patch-rank.vercel.app/api?username=rickstaa&include_all_commits=true)
```

![](https://github-readme-stats-patch-rank.vercel.app/api?username=rickstaa&include_all_commits=true)

```md
![](https://github-readme-stats-patch-rank.vercel.app/api?username=francois-rozet&include_all_commits=true)
```

![](https://github-readme-stats-patch-rank.vercel.app/api?username=francois-rozet&include_all_commits=true)

```md
![](https://github-readme-stats-patch-rank.vercel.app/api?username=FleetAdmiralJakob&include_all_commits=true)
```

![](https://github-readme-stats-patch-rank.vercel.app/api?username=FleetAdmiralJakob&include_all_commits=true)

```md
![](https://github-readme-stats-patch-rank.vercel.app/api?username=meurissemax&include_all_commits=true)
```

![](https://github-readme-stats-patch-rank.vercel.app/api?username=meurissemax&include_all_commits=true)

## count_private=true

```md
![](https://github-readme-stats-patch-rank.vercel.app/api?username=torvalds&count_private=true)
```

![](https://github-readme-stats-patch-rank.vercel.app/api?username=torvalds&count_private=true)

```md
![](https://github-readme-stats-patch-rank.vercel.app/api?username=anuraghazra&count_private=true)
```

![](https://github-readme-stats-patch-rank.vercel.app/api?username=anuraghazra&count_private=true)

```md
![](https://github-readme-stats-patch-rank.vercel.app/api?username=rickstaa&count_private=true)
```

![](https://github-readme-stats-patch-rank.vercel.app/api?username=rickstaa&count_private=true)

```md
![](https://github-readme-stats-patch-rank.vercel.app/api?username=francois-rozet&count_private=true)
```

![](https://github-readme-stats-patch-rank.vercel.app/api?username=francois-rozet&count_private=true)

```md
![](https://github-readme-stats-patch-rank.vercel.app/api?username=FleetAdmiralJakob&count_private=true)
```

![](https://github-readme-stats-patch-rank.vercel.app/api?username=FleetAdmiralJakob&count_private=true)

```md
![](https://github-readme-stats-patch-rank.vercel.app/api?username=meurissemax&count_private=true)
```

![](https://github-readme-stats-patch-rank.vercel.app/api?username=meurissemax&count_private=true)

## count_private=true

```md
![](https://github-readme-stats-patch-rank.vercel.app/api?username=torvalds&include_all_commits=true&count_private=true)
```

![](https://github-readme-stats-patch-rank.vercel.app/api?username=torvalds&include_all_commits=true&count_private=true)

```md
![](https://github-readme-stats-patch-rank.vercel.app/api?username=anuraghazra&include_all_commits=true&count_private=true)
```

![](https://github-readme-stats-patch-rank.vercel.app/api?username=anuraghazra&include_all_commits=true&count_private=true)

```md
![](https://github-readme-stats-patch-rank.vercel.app/api?username=rickstaa&include_all_commits=true&count_private=true)
```

![](https://github-readme-stats-patch-rank.vercel.app/api?username=rickstaa&include_all_commits=true&count_private=true)

```md
![](https://github-readme-stats-patch-rank.vercel.app/api?username=francois-rozet&include_all_commits=true&count_private=true)
```

![](https://github-readme-stats-patch-rank.vercel.app/api?username=francois-rozet&include_all_commits=true&count_private=true)

```md
![](https://github-readme-stats-patch-rank.vercel.app/api?username=FleetAdmiralJakob&include_all_commits=true&count_private=true)
```

![](https://github-readme-stats-patch-rank.vercel.app/api?username=FleetAdmiralJakob&include_all_commits=true&count_private=true)

```md
![](https://github-readme-stats-patch-rank.vercel.app/api?username=meurissemax&include_all_commits=true&count_private=true)
```

![](https://github-readme-stats-patch-rank.vercel.app/api?username=meurissemax&include_all_commits=true&count_private=true)

rickstaa commented:
One moment something seems wrong on my deployment.

francois-rozet commented:
Maybe the target for the number of stars is a bit low. You get a perfect score with ~500 stars which is not uncommon. I'll update that.

francois-rozet commented:
I increased the targets for commits and stars, such that it is now more challenging (and more motivational IMO) to reach the `S` rank. To summarize, a user with 1000 commits per year, 200 PRs, 100 issues, 1000 stars and 100 followers is `S`. 2 times less is `A+`, 4 times less is `A`.

ZjzMisaka commented:
The new system is good, but I'd like to express my personal opinion. I believe that a grading system with only S, A, B tiers is not very good, because people might unconsciously believe that there exist tiers like C, D, F. I've previously stated that a "grading system that satisfies everyone" could ultimately lose its reference value, and the meaning represented by the rank "B" could eventually become the new F.

I trust that people's spirits are not so fragile that they cannot face their own grades, and even lower grades can serve as motivation for personal growth. We don't have to view this system as competitive ranking, instead, we can try to understand this system as a measure of a developer's "growth from the initial state". If we really don't want the grade F to appear, we could use a format like Lv1, Lv2, Lv3... and so forth.

rickstaa commented:
> The new system is good, but I'd like to express my personal opinion. I believe that a grading system with only S, A, B tiers is not very good, because people might unconsciously believe that there exist tiers like C, D, F. I've previously stated that a "grading system that satisfies everyone" could ultimately lose its reference value, and the meaning represented by the rank "B" could eventually become the new F.
> 
> I trust that people's spirits are not so fragile that they cannot face their own grades, and even lower grades can serve as motivation for personal growth. We don't have to view this system as competitive ranking, instead, we can try to understand this system as a measure of a developer's "growth from the initial state". If we really don't want the grade F to appear, we could use a format like Lv1, Lv2, Lv3... and so forth.

@ZjzMisaka thank you so much for giving your two cents 🙏🏻. I agree that the rank names can be improved. As stated in https://github.com/anuraghazra/github-readme-stats/pull/1186#issuecomment-1510409575 however we will tackle this in a subsequent pull request. Possible improvements are found and discussed in https://github.com/anuraghazra/github-readme-stats/issues/2265 👍🏻.

@everybody Feel free to add your two cents. I will merge this PR in:

![Relative date](https://img.shields.io/date/1685098732?style=for-the-badge)

qwerty541 reviewed (commented):
It is very surprising for me that Linus Torvalds, having in his profile a project with an abnormally large number of stars of 150,000+, receives only an A+ grade from the new ranking system. As far as I understand, the new ranking system mainly takes into account the number of open pull requests and issues. Considering that the purpose of this project is to motivate developers and newcomers to contribute more actively in open source projects, then this change looks logical and correct. But I still think that an abnormally large number of stars, commits or followers should also elevate the user to at least the S rank.

ZjzMisaka commented:
> It is very surprising for me that Linus Torvalds, having in his profile a project with an abnormally large number of stars of 150,000+, receives only an A+ grade from the new ranking system. As far as I understand, the new ranking system mainly takes into account the number of open pull requests and issues. Considering that the purpose of this project is to motivate developers and newcomers to contribute more actively in open source projects, then this change looks logical and correct. But I still think that an abnormally large number of stars, commits or followers should also elevate the user to at least the S rank.

I agree with this perspective. It's important to recognize that some users may not be involved in a variety of open-source projects, but they could be focusing on their own open-source projects and making significant contributions there. This type of work should also be encouraged, as it has a positive effect on the open-source community as a whole.

francois-rozet commented:
The ranking takes all stats into account (and mainly stars and prs). To have S or S+, you need to be great everywhere. This is the whole reason for this PR. If you allow to be good in a single stat, then users with 0 stars, 0 issues, 0 PRs, 0 followers, but 100000 commits still get S+, which makes no sense, especially if the goal is to push devs to collaborate. Besides, it is very rare to find other cases like Linus Torvalds with huge number of stars but barely any PRs/issues.

rickstaa commented:
> The ranking takes all stats into account (and mainly stars and prs). To have S or S+, you need to be great everywhere. This is the whole reason for this PR. If you allow to be good in a single stat, then users with 0 stars, 0 issues, 0 PRs, 0 followers, but 100000 commits still get S+, which makes no sense, especially if the goal is to push devs to collaborate. Besides, it is very rare to find other cases like Linus Torvalds with huge number of stars but barely any PRs/issues.

I have to agree with @francois-rozet on this one. An excellent developer `S+` has to score high in all fields. Although the contributions of Linus Torvald are indispensable for the OS community, and he should therefore have the `S+` ranking, we unfortunately cannot deduce this from his GitHub statistics alone 🤔. Furthermore, the big number of commits made by Linus Torvalds that is displayed on the stats card is incorrect. This is caused by a bug in GRS (see https://github.com/anuraghazra/github-readme-stats/issues/564) 😅. The actual commits made by Linus are in the order of 28k, which is not too far from of @anuraghazra, who scores higher on the other stats. 

Although, still would prefer a ranking scheme where I get a `S` ranking instead of a `S+` but I couldn't find weights that don't skew the other developers too much. I therefore still think the solution provided by @francois-rozet is the most balanced one. If you guys can find a combination of weights which gives Linux a `S` ranking without degrading this balance, feel free to open a subsequent pull request. I will likely merge this pull request tomorrow, since it is an enormous improvement compared to the old behaviour, but will consider any improvements in future PRs 👍🏻.

ZjzMisaka commented:
> https://github.com/anuraghazra/github-readme-stats/pull/1186#issuecomment-1562339068

Here's an idea, but this idea may not be mature enough and it may be difficult to implement......
perhaps we can assign individual rankings for commits, PRs, issues, stars, and followers, we can even create a five-dimensional graph based on these data points. This way, we can have a more intuitive visualization of a developer's overall skill level without struggling to evaluate them using just a single score.

rickstaa commented:
> > [#1186 (comment)](https://github.com/anuraghazra/github-readme-stats/pull/1186#issuecomment-1562339068)
> 
> Here's an idea, but this idea may not be mature enough and it may be difficult to implement...... perhaps we can assign individual rankings for commits, PRs, issues, stars, and followers, we can even create a five-dimensional graph based on these data points. This way, we can have a more intuitive visualization of a developer's overall skill level without struggling to evaluate them using just a single score.

Although I agree your idea results in a better developer skill display, I think the extra code and development time required for this idea is not worth it 😅. People can rank developers based on how important they find the individual items displayed on the stats card. The rank was added as a motivator for developers to be well-rounded and OS orientated.

FleetAdmiralJakob commented:
@rickstaa

rickstaa commented:
> @rickstaa

Done 🎉!

rickstaa commented:
@francois-rozet thanks again for implementing this improvement 👍🏻.

sunpm commented:
The weight of commit is very low

rickstaa commented:
> The weight of commit is very low

@sunpm feel free to create an improvement pull request 👍🏻.

lorezyra commented:
[ref:](https://github.com/anuraghazra/github-readme-stats/issues/2265#issuecomment-1566563515) 
> Instead of some letter, why not use what most apps do in gamification: show an XP or numerical level rank? Everyone starts from zero. But few can keep pushing the top.

Dgdiniz commented:
Hi, I think the formula needs some tweeks. I tested here with some values, and for 1000 commits we have:
`{ level: 'B+', score: 69.07599578073207 }`

For 10000 commits we have:
`{ level: 'B+', score: 67.93963214438497 }`

And for 100000 commits we have:
`{ level: 'B+', score: 67.93963214436843 }`

So the values are converging too fast, and after a while makes no difference to increase a particular value.

francois-rozet commented:
This is expected. People with 1000 commits per year are already in the top 5 percentiles (for that statistic). So more commits will not improve the total rank dramatically. If you want to improve the rank, you will have to work on the other stats.

rickstaa commented:
> Hi, I think the formula needs some tweeks. I tested here with some values, and for 1000 commits we have: `{ level: 'B+', score: 69.07599578073207 }`
> 
> For 10000 commits we have: `{ level: 'B+', score: 67.93963214438497 }`
> 
> And for 100000 commits we have: `{ level: 'B+', score: 67.93963214436843 }`
> 
> So the values are converging too fast, and after a while makes no difference to increase a particular value.

I agree that in the current formula levels out too quickly when people are only good in one statistic. Luckily, @francois-rozet already created a new PR that in my opinion offers a good balance 🙏🏻. @Dgdiniz maybe you can review https://github.com/anuraghazra/github-readme-stats/pull/2762#pullrequestreview-1457301484? I hosted it on https://github-readme-stats-git-rank-rickstaa.vercel.app:

```md
[![Anurag's GitHub stats](https://github-readme-stats-git-rank-rickstaa.vercel.app/api?username=Dgdiniz)](https://github.com/anuraghazra/github-readme-stats)
```

[Anurag's GitHub stats](https://github-readme-stats-git-rank-rickstaa.vercel.app/api?username=Dgdiniz)](https://github.com/anuraghazra/github-readme-stats)

Dgdiniz commented:
Hi @rickstaa, I tested here but the behavior is similar. After a while makes no difference improving a category. I was thinking in a simpler approach. Each category has a weight, so a commit is 2 points and a star is 4 points. So each rank just needs a certain number of points. So each category could reach rank S. The number of points could be calculated using some top contributor as reference. Like 200k points is S+. The points destribution could follow some curve, so from C+ to B- we need few points, but from A+ to S- we need much more points. It's very unlikely to have 50k commits, but some repos have thousands of stars, so I don't know the best way to weight the categories. But today we can have billions of points in a category and still be at rank B. But the idea of forcing points in more than one category is also cool.

rickstaa commented:
> Hi @rickstaa, I tested here but the behavior is similar. After a while makes no difference improving a category. I was thinking in a simpler approach. Each category has a weight, so a commit is 2 points and a star is 4 points. So each rank just needs a certain number of points. So each category could reach rank S. The number of points could be calculated using some top contributor as reference. Like 200k points is S+. The points destribution could follow some curve, so from C+ to B- we need few points, but from A+ to S- we need much more points. It's very unlikely to have 50k commits, but some repos have thousands of stars, so I don't know the best way to weight the categories. But today we can have billions of points in a category and still be at rank B. But the idea of forcing points in more than one category is also cool.

@Dgdiniz thanks for your feedback. Although I like the simplicity of your system, I love that the current system promotes good code practices like working with pull requests and being a well-rounded developer 🤔. As can be seen in https://github.com/anuraghazra/github-readme-stats/pull/2762#issuecomment-1573902435 I think all ranks are douable when people: make use of both pull requests and commits and get stars and followers by creating OS projects that have value to other people.

rickstaa commented:
However, as said before, this is a community-owned OS project, so all feedback is welcome. Therefore feel free to open a feature request with your proposal. If enough people upvote your idea, I will review it (see https://github.com/anuraghazra/github-readme-stats/issues/1935).


#1186 and #2762 were/will be merged because there was a lot of demand for a better ranking system (see https://github.com/anuraghazra/github-readme-stats/issues/455). @francois-rozet did a fantastic job at improving the old system since getting a rank higher or lower than an A+ in the old system was tough. I also think that creating a rank system that will make everybody happy is not doable since we don't have the ground truth of what is an average GitHub developer 😅.

aaron-muti-420 reviewed (commented):
Con you consider leaving comments
